pretyping/coercion.ml

```diff
@@ -622,6 +622,29 @@ let lookup_reversible_path_to_common_point env sigma ~src_expected ~src_inferred
   in
     aux r
 
+type hook = env -> evar_map -> flags:Evarconv.unify_flags -> constr ->
+  inferred:types -> expected:types -> (evar_map * constr) option
+type hook_rec = { hr_id : string; hr_active : bool; hr_hook : hook }
+
+let active_hooks = ref ([] : hook list)
```

Comment:
I would rather do something like
```suggestion
let all_hooks = ref (CString.Map.empty : hook CString.Map.t)

let active_hooks = Summary.ref ~name:"coercion hooks" ([] : string list)

let register_hook name h = assert (not (CString.Map.mem name !all_hooks)); all_hooks := CString.Map.add name h !all_hooks

(* maybe also remove "name" from the tail of the list so it appears uniquely *)
let activate_hook name = assert (CString.Map.mem name !all_hooks); active_hooks := name :: !active_hooks

(* could also have deactivate_hook *)

let active_hooks () = List.map (fun name -> CString.Map.get name !all_hooks) !active_hooks
```

Reply:
My goal was to make coercion fast even if unfreeze becomes a bit slower. But maybe I was overengineering it so let's go for your simpler impleme which seems to be what everyone is expecting.